### PR TITLE
Add new differentiation method based on rewinding the tape [original PR, not to be merged]

### PR DESCRIPTION
**Pull request by trbromley**
WIP

### CODE REVIEW

**Comment by trbromley** [comment]
[ch4083]

File: pennylane/tape/tapes/rewind.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                raise ValueError(
+                    f"The {m.return_type.value} return type is not supported with the rewind gradient method"
+                )
+
+        method = options.get("method", "analytic")
+
+        if method == "device":
+            # Using device mode; simply query the device for the Jacobian
+            return self.device_pd(device, params=params, **options)
+        elif method == "numeric":
```
Comment (albi3ro) [inline pennylane/tape/tapes/rewind.py:54]: If it's numeric we can just send it to the Jacobian tape's execution `super().jacobian`.  Defeats the purpose of using a rewind tape in the first place, but if they really want to...

File: pennylane/tape/tapes/rewind.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        if not device.capabilities().get("returns_state") or isinstance(device, DefaultMixed) \
+            or not hasattr(device, "_apply_operation"):
+            # TODO: consider renaming returns_state to, e.g., uses_statevector
+            # TODO: consider adding a capability for mixed/pure state
+            # TODO: consider adding capability for apply_operation
+            raise qml.QuantumFunctionError("The rewind gradient method is only supported on statevector-based devices")
+
+        # Perform the forward pass
+        # TODO: Could we use lower-level like device.apply, since we just need the state?
+        self.execute(device, params=params)
```
Comment (albi3ro) [inline pennylane/tape/tapes/rewind.py:66]: Would be wonderful if we could use a memoized/ cached circuit execution here.  Though I guess execution memorization is on the to-do list for all the tapes.

File: pennylane/tape/tapes/rewind.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        phi = device._state  # TODO: Do we need dev._state or dev.state?
+
+        lambdas = [device._apply_operation(phi, obs) for obs in self.observables]
+
+        jac = np.zeros((len(lambdas), len(self.trainable_params)))
```
Comment (albi3ro) [inline pennylane/tape/tapes/rewind.py:71]:
```suggestion
        jac = np.zeros((len(self.observables), len(self.trainable_params)))
```

Same thing, but it makes clearer what the two dimensions represent.

File: pennylane/tape/tapes/rewind.py (hunk 1 above)
Comment (trbromley) [reply pennylane/tape/tapes/rewind.py:54]: Thanks, ok I made that change

File: pennylane/tape/tapes/rewind.py (hunk 2 above)
Comment (trbromley) [reply pennylane/tape/tapes/rewind.py:66]: There is device-side [caching](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.devices.default_qubit.DefaultQubit.html) - but there's also lots going in between `tape.execute()` and `device.execute()` which may be unnecessary. Also, we only need to update the state, so even `dev.apply()` might be enough.

**Review by trbromley — COMMENTED** [review]
Thanks @albi3ro

**Comment by trbromley** [comment]
Closing this since it is superseded by https://github.com/PennyLaneAI/pennylane/pull/1032

**Pull request closed without merging**